[mkcal] Reset alarms for recurring events when an exception is removed.

dcaliste opened the pull request:
@pvuorela , yet another alarm issue for recurring events with exceptions. I wrongly assumed that removing an exception in the QML bindings always add an EXDATE for that occurrence. So I initially designed the tests with this assumption in mind in mkcal. But that's not true (anymore?) : for instance, create a recurring event, create an exception by changing the start time for instance on one occurrence, and then delete the exception. The normal occurrence is back, because no EXDATE is set. Delete the normal occurrence and an EXDATE is set.

As you can see from the old test here, the exception deletion and the EXDATE addition was executed at the same time, while if you run the test with two separate actions, and testing the alarm value in-between (as newly implemented), you observe that the alarm is assuming the EXDATE when the exception is deleted and the test is failing.

That's because how the alarms are implemented : there is one alarm for the parent (taken into account the exceptions are EXDATEs) and one alarm for each exception. Removing one exception without assuming that it is replaced by an EXDATE thus require to reset the alarm of the parent. That's what the current patch is doing.

pvuorela commented on src/extendedstorage.cpp:
```diff
@@ -738,11 +735,28 @@ void ExtendedStorage::Private::clearAlarms(const Incidence::Ptr &incidence)
+            }
+        }
     }
+    setAlarms(recurring, calendar);
```
This somehow bothers me a bit when clearAlarms() sets alarms. Not that I have alternatives to suggest.

dcaliste replied:
I'm pushing an additionnal commit with a proposition to remove the clearAlarms(list, calendar) function and put it where it is used since it was used only in one place.

pvuorela replied:
Hey that's very nice. The code suddenly became a lot easier to follow. Could you still squash that commit to the first one.

One thing that was also confusing me was the overloaded methods that modified either a single instance, a list or a notebook. That's now almost all gone. For the remaining clearAlarms(notebook_uid) could consider would it be more readable as clearNotebookAlarms() or something, but up to you, just a related suggestion.

dcaliste replied:
I've squashed the commits and renamed the `clearAlarms(notebookUid)` into `clearAlarmsByNotebook(notebookUid)`.

pvuorela commented:
Another thing here is that by removing an exception from the UI it's confusing when the originally replaced occurrence comes back. Thinking we should somewhere on upper level add an exception, the latest when the exception is removed.

But since remote sync can in theory at least delete an exception and bring back the original occurrence, suppose this change per se is fine and we could merge.

Makes sense?

dcaliste commented:
> Another thing here is that by removing an exception from the UI it's confusing when the originally replaced occurrence comes back. Thinking we should somewhere on upper level add an exception, the latest when the exception is removed.

That's a good point. I hesitated to bring back the original behaviour in nemo-qml-plugin-calendar, by adding an exception to the parent when deleting an exception. Since you go also in the same direction, I'm going to propose a PR for it.

Having the pattern (parent modified + exception deleted) in the same database update, made me rethink a bit the change in this PR. I've modified a bit the `update()`routine to join up the modified and deleted cases more.

pvuorela reviewed (approved):
Alright, let's have this for starters.
